compiler/rustc_middle/src/lint.rs

```diff
@@ -398,8 +398,25 @@ pub fn lint_level(
             }
         }
 
-        // Finally, run `decorate`.
-        decorate(&mut err);
+        // Finally, run `decorate`. `decorate` can call `trimmed_path_str` (directly or indirectly),
+        // so we need to make sure when we do call `decorate` that the diagnostic is eventually
+        // emitted or we'll get a `must_produce_diag` ICE.
+        //
+        // When is a diagnostic *eventually* emitted? Well, that is determined by 2 factors:
+        // 1. If the corresponding `rustc_errors::Level` is beyond warning, i.e. `ForceWarning(_)`
+        //    or `Error`, then the diagnostic will be emitted regardless of CLI options.
+        // 2. If the corresponding `rustc_errors::Level` is warning, then that can be affected by
+        //    `-A warnings` or `--cap-lints=xxx` on the command line. In which case, the diagnostic
+        //    will be emitted if `can_emit_warnings` is true.
+        {
+            use rustc_errors::Level as ELevel;
+            if matches!(err_level, ELevel::ForceWarning(_) | ELevel::Error)
+                || sess.dcx().can_emit_warnings()
```

Comment:
This check seems to pass for the UI test with including the `-Dunused_must_use` case, but I'm still not very confident in this check.

Reply:
I think it would be best if we instead used an if statement to consider when to skip calling decorate.

Something like

```rust
let skip = err_level == rustc_errors::Warning && !sess.dcx().can_emit_warnings();
if !skip {
    decorate(&mut err);
}
```

Which would have better maintainability.